Add timeout values to VA RoundTripper.

jsha opened the pull request:
Fixes #3868.

rolandshoemaker commented on va/va.go:
```diff
@@ -393,6 +393,11 @@ func (va *ValidationAuthorityImpl) fetchHTTP(ctx context.Context, identifier cor
 		// select.
 		DialContext: dialer.DialContext,
+		// These match the values from DefaultRoundTripper in net/http.
+		MaxIdleConns:          100,
```
Given we want to use a new connection each time, and we explicitly disable KeepAlives, don't we want this to be 0?

rolandshoemaker replied:
And if so do we need a `IdleConnTimeout` set at all?

rolandshoemaker commented on va/va.go:
```diff
@@ -393,6 +393,11 @@ func (va *ValidationAuthorityImpl) fetchHTTP(ctx context.Context, identifier cor
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
```
From the docs:
> // ExpectContinueTimeout, if non-zero, specifies the amount of
    // time to wait for a server's first response headers after fully
    // writing the request headers if the request has an
    // "Expect: 100-continue" header. Zero means no timeout and
    // causes the body to be sent immediately, without
    // waiting for the server to approve.
    // This time does not include the time to send the request header.

We don't send the `Expect: 100-continue` so it sounds like we have no reason to set this.

cpu replied:
Counter-point: setting it seems pretty harmless and if some confluence of stupid circumstances down the road results in this being relevant we wouldn't want no timeout.

rolandshoemaker replied:
Counter-counter-point: this change is being made to fix a problem we are having now and have (presumably) pinpointed to a missing handshake timeout. Adding in a possible-preventative fix that we don't actually know what will do in the internals is just going to muddy the waters for the real fix that we know we need.

cpu replied:
@rolandshoemaker That's fair. I don't feel that strongly about it.

@jsha Do you have a preference?

jsha commented:
Added a comment: For lots of these values, 0 means "unlimited," so I had originally copied the nonzero default values from net/http. I lowered the values for idle connections, and removed the ExpectContinueTimeout. We don't use ExpectContinue so it doesn't matter, but a 0 for ExpectContinueTimeout means sending the full request right away.

rolandshoemaker commented on va/va.go:
```diff
@@ -393,6 +393,10 @@ func (va *ValidationAuthorityImpl) fetchHTTP(ctx context.Context, identifier cor
 		DialContext: dialer.DialContext,
+		// We don't want idle connections, but 0 means "unlimited," so we pick 1.
+		MaxIdleConns:        1,
+		IdleConnTimeout:     1 * time.Second,
```
(also side note: the `1 *` here is redundant)

jsha replied:
I think there's not really a benefit to that, and I worry slightly about tickling weird races (we had a hard to track test flake a while ago using millisecond timeouts). We don't have an issue right now with idle connections that are actually tracked (the TLS handshaking connections just kinda fall off the map), so I'm not concerned with making this the smallest possible number. I did fix the `1 * ` issue though.

rolandshoemaker commented:
Also perhaps this just requires a comment clarifying in the code but it's slightly confusing that we set `DisableKeepAlives` above and then set `MaxIdleConns` to _anything_. Having it set to 0 (or unlimited) seems fine if we already explicitly disabled the reuse of connections anyway.

jsha commented:
```
    // DisableKeepAlives, if true, prevents re-use of TCP connections
    // between different HTTP requests.
```

This doesn't actually say it will close idle connections immediately. I think it's reasonable to have both DisableKeepAlives and a small MaxIdleConnections.

rolandshoemaker commented:
I think this is just poor documentation of `DisableKeepAlives` based on reading the transport code, that said I'm not that concerned about this, ¯\_(ツ)_/¯.
